####################################################################################################
# This scripts takes as inputs the results by ANNOVAR generated by LACEpipeline.sh 
# The scripts: 1) remove any mutations with synonymous or unknown function 
#              2) remove any mutations with less then thr_alleles_ratio reads of support 
#              3) remove any mutations with MAF greater than thr_maf (likely SNPs) 
#              4) remove any mutations with occurence less than thr_freq in all the 4 time points 
####################################################################################################

# set working directory
setwd("/data_directory/BimiB/share/LACE/MELANOMA/annotated/")

# load required libraries
library("biomaRt")

# settings
thr_alleles_ratio <- 2
thr_maf <- 0.01
thr_freq <- 0.05

# get list of files to be considered and cells information
listSC <- dir(pattern=".exonic_variant_function",ignore.case=TRUE)
cellInfo <- readRDS("../data_info.rds")

# create data structure to save the results
tbFinal <- setNames(data.frame(matrix(ncol=12,nrow=0)), c("scID","Time","Gene","Chr","PosStart","PosEnd","REF","ALT","MutType","rs_ID","depth","Allele_Ratio"))
cn <- colnames(tbFinal)

# consider each cell
cont <- 0
for(sc in listSC) {
    
    # get current cell name
    run <- strsplit(x = sc, split = "\\.")[[1]][1]
    
    # read vcf annotated by ANNOVAR for current cell
    tmpTbl <- read.table(file = sc, header = FALSE, sep = "\t", stringsAsFactors = FALSE)
    tmpTbl$scID <- run
    
    # get information about time of experiment
    tmpTime <- as.character(cellInfo$Age[cellInfo$Run==run])
    tmpTbl$Time <- tmpTime
    
    # get genes names
    tmpTbl$V3 <- sapply(as.character(tmpTbl$V3), function(x){strsplit(x, ":")[[1]][1]})
    
    # get the remaining information
    OtherInfo <- tmpTbl[, "V18"]
    tmpTbl <- tmpTbl[, c("scID", "Time", "V3", "V4", "V5", "V6", "V7", "V8", "V2", "V11")]
    colnames(tmpTbl) <- c("sc_ID", "Time", "Gene", "Chr", "PStart", "PEnd", "Ref", "Alt", "Mut_type", "RS_id")
    
    # extract quality and coverage information
    for(m in 1:length(OtherInfo)) {
        
        info <- unlist(strsplit(OtherInfo[m],":"))
        
        if(length(info) < 5) {
            depth <- NA
            RA <- NA
        }
        else {
            RefAlt <- unlist(strsplit(info[2], ","))
            depth <- sum(as.numeric(RefAlt))
            if(depth == 0) {
                RA <- NA
                }
            else {
                RA <- as.numeric(RefAlt[2])
            }
        }
        tmpTbl$depth[m] <- depth
        tmpTbl$Allele_Ratio[m] <- RA
        
    }
    
    # save results for current cell
    colnames(tmpTbl) <- cn
    tbFinal <- rbind(tbFinal,tmpTbl)
    
    cont = cont + 1
    cat(paste0(cont, " done | ", round(cont/length(listSC),digits = 2),"%\r"))
    
}

tbFinal$UIDsnp <- paste0(tbFinal$Gene,"_",tbFinal$Chr,"_",tbFinal$PosStart,ifelse(tbFinal$rs_ID == '.', "", paste0("_",tbFinal$rs_ID)))
saveRDS(tbFinal,file=paste0("../cells_aggregate_info.rds"))

tbFinal <- readRDS(file=paste0("../cells_aggregate_info.rds"))

# remove synonymous mutations and the ones marked as unknown
tbFinal <- tbFinal[tbFinal$MutType!="synonymous SNV"& tbFinal$MutType!="unknown",]

u_rsID <- as.character(unique(tbFinal$rs_ID[startsWith(as.character(tbFinal$rs_ID),prefix="rs")]))

ensembl_snp = useMart("ENSEMBL_MART_SNP",dataset="hsapiens_snp")
SNPInfo <- getBM(attributes=c("refsnp_id",
                              "minor_allele",
                              "minor_allele_freq",
                              "polyphen_prediction",
                              "sift_prediction",
                              "clinical_significance"),
                 filters = "snp_filter",
                 values = u_rsID,
                 mart = ensembl_snp)

saveRDS(SNPInfo,file=paste0("../SNPInfo.rds"))

scMutInfo <- merge.data.frame(tbFinal, SNPInfo, by.x = "rs_ID", by.y = "refsnp_id", all.x = TRUE, all.y = TRUE)
saveRDS(scMutInfo,file=paste0("../scMutInfo.rds"))


###########################################################################################################
### START FILTERING STEP ###
###########################################################################################################

scMutInfo <- readRDS(file=paste0("../scMutInfo.rds"))

scMutInfo <- scMutInfo[((!is.na(scMutInfo$Allele_Ratio)) & (scMutInfo$Allele_Ratio >= thr_alleles_ratio)),]

mutWOrs <- scMutInfo[scMutInfo$rs_ID == ".", ]
mutWrs <- scMutInfo[scMutInfo$rs_ID != ".", ]

FilterMutations <- function(sc_r) {
    
    # if not a snp, keep it
    if(sc_r["rs_ID"] == ".") {
        return(TRUE)
    }
    
    # filter using MAF
    if(sc_r["minor_allele"] == sc_r["ALT"] && sc_r["minor_allele_freq"] > thr_maf) {
        return(FALSE)
    }
    else if(sc_r["minor_allele"] == sc_r["REF"] && sc_r["minor_allele_freq"] < (1-thr_maf)) {
        return(FALSE)
    }
    
    return(TRUE)
    
}

mutWrs_filt <- mutWrs[apply(mutWrs, 1, function(x) FilterMutations(x)),]
snpMut_filt <- rbind(mutWOrs, mutWrs_filt)
saveRDS(snpMut_filt,file=paste0("../snpMut_filt.rds"))


###########################################################################################################
### MAKE CELLS/MUTATIONS MATRIX ###
###########################################################################################################

snpMut_filt <- readRDS(file=paste0("../snpMut_filt.rds"))

# consider only cells passing quality checks
load("valid_clones_mapping.RData")
snpMut_filt = snpMut_filt[which(snpMut_filt$scID%in%valid_clones_mapping$Run),]

tblscMut_T1 <- table(snpMut_filt$scID[snpMut_filt$Time == "before treatment"], snpMut_filt$UIDsnp[snpMut_filt$Time == "before treatment"])
tblscMut_T2 <- table(snpMut_filt$scID[snpMut_filt$Time == "4d on treatment"], snpMut_filt$UIDsnp[snpMut_filt$Time == "4d on treatment"])
tblscMut_T3 <- table(snpMut_filt$scID[snpMut_filt$Time == "28d on treatment"], snpMut_filt$UIDsnp[snpMut_filt$Time == "28d on treatment"])
tblscMut_T4 <- table(snpMut_filt$scID[snpMut_filt$Time == "57d on treatment"], snpMut_filt$UIDsnp[snpMut_filt$Time == "57d on treatment"])

# binarization
tblscMut_T1[which(tblscMut_T1 > 0)] <- 1
tblscMut_T2[which(tblscMut_T2 > 0)] <- 1
tblscMut_T3[which(tblscMut_T3 > 0)] <- 1
tblscMut_T4[which(tblscMut_T4 > 0)] <- 1

Mut_keep = c()

cs <- colSums(tblscMut_T1, na.rm = TRUE) / nrow(tblscMut_T1)
Mut_keep = c(Mut_keep, colnames(tblscMut_T1)[cs > thr_freq])
cs <- colSums(tblscMut_T2, na.rm = TRUE) / nrow(tblscMut_T2)
Mut_keep = c(Mut_keep, colnames(tblscMut_T2)[cs > thr_freq])
cs <- colSums(tblscMut_T3, na.rm = TRUE) / nrow(tblscMut_T3)
Mut_keep = c(Mut_keep, colnames(tblscMut_T3)[cs > thr_freq])
cs <- colSums(tblscMut_T4, na.rm = TRUE) / nrow(tblscMut_T4)
Mut_keep = c(Mut_keep, colnames(tblscMut_T4)[cs > thr_freq])

snpMut_filt_freq <- snpMut_filt[snpMut_filt$UIDsnp %in% Mut_keep,]

saveRDS(snpMut_filt_freq,file=paste0("../snpMut_filt_freq.rds"))
